tokenizers/src/tokenizer/added_vocabulary.rs

```diff
@@ -90,6 +91,37 @@ impl std::cmp::Eq for AddedToken {}
 
 type MatchingSet = (AhoCorasick, Vec<u32>);
 
+lazy_static! {
+    static ref STARTS_WITH_WORD: Regex = Regex::new(r"^\w").unwrap();
+    static ref ENDS_WITH_WORD: Regex = Regex::new(r"\w$").unwrap();
+    static ref RIGHTMOST_SPACE_AT_START: Regex = Regex::new(r"^\s*").unwrap();
+    static ref LEFTMOST_SPACE_AT_END: Regex = Regex::new(r"\s*$").unwrap();
+}
+
+fn ends_with_word(sentence: &str) -> bool {
+    ENDS_WITH_WORD.is_match(sentence)
+}
+
+fn starts_with_word(sentence: &str) -> bool {
+    STARTS_WITH_WORD.is_match(sentence)
+}
+
+fn space_leftmost_at_end(sentence: &str) -> usize {
+    if let Some(match_) = LEFTMOST_SPACE_AT_END.find(sentence) {
+        match_.start()
+    } else {
+        // This should never happen since the Regex should match all the time
+        sentence.len()
+    }
+}
+fn space_rightmost_at_start(sentence: &str) -> usize {
+    if let Some(match_) = RIGHTMOST_SPACE_AT_START.find(sentence) {
+        match_.end()
+    } else {
+        // This should never happen since the Regex should match all the time
+        0
+    }
+}
```

Comment:
Wouldn't it be better to return a `Result` here ?

Reply:
Well if the regexp doesn't match, it's strictly equivalent to matching nothing, right ?

I do that to avoid adding and `unwrap` which could potential `panic!` which makes users pretty unhappy.

I don't think the pathway should be taken, but if it ever is, then I think `0` is a valid default which prevents panicking and is still, correct. Wdyt ?

Reply:
I understand adding `Result`s all the way up to the public API is cumbersome but to me is the cleaner approach : notify the user there was an error / undefined / unexpected behavior and let him handle it.

Returning 0 is hiding the fact that you've run into an unexpected case/behavior (aka silent error). Adding comments to clarify code is a smell imo, make it comprehensible in the first place, you almost always can !

But then again, it's more work.

Reply:
> I understand adding Results all the way up to the public API is cumbersome but to me is the cleaner approach : notify the user there was an error / undefined / unexpected behavior and let him handle it.

In that case it's not true. Finding 0 match is different than an error.
I could very well change the regex to be `\s+` and then the `else` branch would be expected and working as intended.

Is adding a `warn!("AddedToken with `single_word` seems to have an issue, Please report this")` in that code path OK ?
It it warning the user, but still prevent catastrophing failure ?

We don't expect the code path to be taken, but it's not preventing the algorithm from working.

Reply:
Am I correct in assuming that if we go through the `else` clause, it means the user has inputted a malformatted `.lstrip` or `.rstrip` token ?

Reply:
No, the else clause will trigger, when `REGEXP.find()` returns `None`.

This is normal when the REGEX fails to find any match in the submitted string.
This particular brand of REGEX should match all the time (since it should match the empty string being `^\s*` and `\s*$`).
So I don't expect it to not match. But if it doesn't match, it's roughly the same as saying no space where found.

Reply:
But that's what I don't understand, shouldn't there be a space ?

Reply:
Oh no no necessarily.

`AddedTokens("<mask>", lstrip=True)` means you want it to match "Something <mask>  else" -> ['Something", " <mask>", " else"]. Meaning you are actually capturing the left spaces in addition to your token (effectively tripping it from being seen for your model).

But "Something<mask>else" will also capture ["Something", "<mask>", "else"]. no space have been deleted.
If you want this to NOT capture you need to activate `single_word=True` too.

But both options are orthogonal and don´t  necessarily need one another (although I think in practice they are probably reasonned about in conjunction)

Reply:
Ok that makes sense. One thing tho : I would remove the comment, imho it's misleading !

Reply:
@McPatate I merged since I am starting the necessary work to release this before transformers's own release on thursday, but we can continue the discussion to make this clear.